lib/hex_web/store/s3.ex

```diff
@@ -90,33 +87,15 @@ defmodule HexWeb.Store.S3 do
     HexWeb.Plug.redirect(conn, url)
   end
 
-  def upload(bucket, path, headers, data) do
+  def upload(bucket, path, data, opts \\ []) do
+    opts = Keyword.put(opts, :acl, :public_read)
     # TODO: cache
-    bucket     = Application.get_env(:hex_web, bucket)
-    S3.put_object(bucket, path, data, headers)
-    S3.put_object_acl(bucket, path, %{acl: :public_read})
+    Application.get_env(:hex_web, bucket)
+    |> S3.put_object(path, data, opts)
   end
 
   defp list(bucket, prefix) do
-    bucket = Application.get_env(:hex_web, bucket)
-    list_all(bucket, prefix, nil, %{})
-  end
-
-  defp list_all(bucket, prefix, marker, keys) do
-    opts = %{"prefix" => prefix}
-    if marker do
-      opts = Dict.put(opts, "marker", marker)
-    end
-
-    result = S3.list_objects(bucket, opts)
-
-    new_keys = result[:contents]
-    all_keys = new_keys ++ keys
-
-    if result[:is_truncated] do
-      list_all(bucket, prefix, List.last(new_keys), all_keys)
-    else
-      all_keys
-    end
+    Application.get_env(:hex_web, bucket)
+    |> S3.stream_objects(prefix: prefix)
```

Comment:
This is much simpler now. Will need to see if the maps that this streams are compatible.